tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java

```diff
@@ -0,0 +1,304 @@
+
+public class LossesImpl {
+
+  /**
+   * Squeeze or expand last dimension if needed with a sampleWeights of one.
+   *
+   * <ol type="1">
+   *   <li>Squeezes last dim of <code>predictions</code> or <code>labels</code> if their rank differs by 1 (using
+   *       {@link #removeSqueezableDimensions}).
+   *   <li>Squeezes or expands last dim of <code>sampleWeight` if its rank differs by 1 from the new
+   *       rank of <code>predictions`. If <code>sampleWeight` is scalar, it is kept scalar./li>
+   * </ol>
+   *
+   * @param tf the TensorFlow Ops
+   * @param predictions Predicted values, a <code>Operand</code> of arbitrary dimensions.
+   * @param labels Optional label <code>Operand</code> whose dimensions match <code>prediction</code>.
+   * @return Tuple of <code>prediction</code>, <code>label</code> and <code>sampleWeight</code>. Each of them possibly has the last
+   *     dimension squeezed, <code>sampleWeight</code> could be extended by one dimension. If <code>sampleWeight</code>
+   *     is null, (prediction, label) is returned.
+   */
+  public static <T extends TNumber> Tuple<T> squeezeOrExpandDimensions(
+      Ops tf, Operand<T> labels, Operand<T> predictions) {
+    return squeezeOrExpandDimensions(tf, labels, predictions, null);
+  }
+
+  /**
+   * Squeeze or expand last dimension if needed.
+   *
+   * <ol type="1">
+   *   <li>Squeezes last dim of `predictions` or `labels` if their rank differs by 1 (using *
+   *       `confusion_matrix.remove_squeezable_dimensions`). *
+   *   <li>Squeezes or expands last dim of `sampleWeight` if its rank differs by 1 from the new *
+   *       rank of `predictions`. If `sampleWeight` is scalar, it is kept scalar./li> *
+   * </ol>
+   *
+   * @param tf the TensorFlow Ops
+   * @param predictions Predicted values, a <code>Operand</code> of arbitrary dimensions.
+   * @param labels Optional label <code>Operand</code> whose dimensions match <code>prediction
+   *     </code>.
+   * @param sampleWeight Optional sample weight(s) <code>Operand</code> whose dimensions match<code>
+   *     prediction</code>.
+   * @return Tuple of <code>prediction</code>, <code>label</code> and <code>sampleWeight</code>.
+   *     Each of them possibly has the last dimension squeezed, <code>sampleWeight</code> could be
+   *     extended by one dimension. If <code>sampleWeight</code> is null, (prediction, label) is
+   *     returned.
+   */
+  public static <T extends TNumber> Tuple<T> squeezeOrExpandDimensions(
+      Ops tf, Operand<T> labels, Operand<T> predictions, Operand<T> sampleWeight) {
+    Tuple<T> tuple = new Tuple<>(labels, predictions);
+
+    Shape predictionsShape = predictions.asOutput().shape();
+    long predictionsRank = predictionsShape.numDimensions();
+
+    if (labels != null) {
+      Shape labelsShape = labels.asOutput().shape();
+      long labelRank = labelsShape.numDimensions();
+      if (labelRank != Shape.UNKNOWN_SIZE && predictionsRank != Shape.UNKNOWN_SIZE) {
```

Comment:
This logic is checking to see if both objects ranks are known (not` Shape.unknown()`). If both ranks are known, then it checks to see if the shapes are already in the right relationship or not. If not in the right relationship, then call `removeSqueezableDimensions`. It is basically an optimization to avoid doing the work in `removeSqueezableDimensions` if it does not need to be done.